src/goto-programs/goto_cfg.cpp

```diff
@@ -165,3 +194,646 @@ void goto_cfg::dump_graph() const
+{
+  std::unordered_map<CFGNode, size_t> levels;
+  levels[root] = 0;
+
+  std::set<CFGNode> worklist({root});
+  while (!worklist.empty())
+  {
+    CFGNode current = *worklist.begin();
+    worklist.erase(current);
+    if (!edges.count(current))
+      continue;
+
+    size_t level = levels[current] + 1;
+    for (const CFGNode &nodes : edges.at(current))
+    {
+      levels[nodes] = level;
+      worklist.insert(nodes);
+    }
+  }
+  return levels;
+}
+
+std::unordered_set<CFGNode>
+Dominator::DomTree::get_subtree(const CFGNode &n) const
+{
+  std::unordered_set<CFGNode> nodes({n});
+  std::set<CFGNode> worklist({n});
+
+  while (!worklist.empty())
+  {
+    const CFGNode current = *worklist.begin();
+    worklist.erase(current);
+
+    if (!edges.count(current))
+      continue;
+
+    for (const auto &inner : edges.at(current))
+    {
+      worklist.insert(inner);
+      nodes.insert(inner);
+    }
+  }
+
+  return nodes;
+}
+
+Dominator::DomTree::DomTree(const Dominator &dom) : root(dom.start)
+{
+  goto_cfg::foreach_bb(dom.start, [this, &dom](const CFGNode &n) {
+    if (n == root)
+      return;
+
+    CFGNode root = dom.idom(n);
+    auto [val, ins] = edges.insert({root, std::unordered_set<CFGNode>()});
+    val->second.insert(n);
+  });
+}
+
+void ssa_promotion::promote()
+{
+  // TODO: The promote_node algorithm works by:
+  // 1. Identify and compute the liveness of each var (read-only)
+  // 2. Insert the phi nodes (affects all instructions)
+  // 3. Rename the phi nodes (affects per var instructions)
+  // We could process 1 and 3 in parallel
+  for (auto &[k, v] : cfg.basic_blocks)
+  {
+    if (_skip.count(k))
+      continue;
+
+    assert(v.size());
+    promote_node(goto_functions.function_map[k].body, v[0]);
+  }
+  goto_functions.update();
+}
+
+void replace_symbols_in_expr(
+  expr2tc &use,
+  symbolt *symbol,
+  unsigned last_id_in_bb,
+  const irep_idt var_name)
+{
+  if (!use)
+    return;
+
+  if (is_phi2t(use))
+    return;
+
+  if (is_symbol2t(use) && to_symbol2t(use).thename == var_name)
+  {
+    use = symbol2tc(use->type, symbol->id);
+    return;
+  }
+
+  use->Foreach_operand([symbol, last_id_in_bb, var_name](expr2tc &e) {
+    replace_symbols_in_expr(e, symbol, last_id_in_bb, var_name);
+  });
+}
+
+std::unordered_map<std::string, ssa_promotion::symbol_information>
+ssa_promotion::extract_symbol_information(const CFGNode &start) const
+{
+  std::unordered_map<std::string, symbol_information> symbol_map;
+  std::unordered_set<std::string> unpromotable_symbols;
+
+  const auto insert_definition = [&symbol_map](
+                                   const irep_idt &var,
+                                   goto_programt::instructionst::iterator &I,
+                                   const CFGNode &bb) {
+    // Are we dealing with a global?
+    if (!symbol_map.count(var.as_string()))
+      return;
+    ;
+    symbol_information &sym = symbol_map[var.as_string()];
+
+    sym.def_blocks.insert(bb);
```

Comment:
Shall we consider using `emplace` when inserting elements into containers such as `std::unordered_set` to avoid unnecessary copying or moving?